node/node: increase test coverage and other improvements

tbjump opened the pull request:
Best to review commit-by-commit. They are all related, hence in the same PR but if you'd like me to split them up into separate PRs that can easily be done. Just let me know.

tbjump commented:
Current dependencies on/for this PR:
* main
  * **PR #3055** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3055" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #3024** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3024" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #3067** <a href="https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3067" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/wormhole-foundation/wormhole/3067?utm_source=stack-comment).

tbjump commented:
node/pkg/node.go has now 79% test coverage. 

```
go test -v -ldflags '-extldflags "-Wl,--allow-multiple-definition" ' -coverprofile=coverage.out
go tool cover -html=coverage.out
```

tbjump commented:
Here is a summary of golang test coverage improvements on this PR (d4fedcf767df7c5de341fa5a7ac1ea6d031623fc) vs 6ad0bbf85b41f3bc1ddf4fd4188b8a1eb1744cbe

/node/pkg/processor/broadcast.go (0.0%) --> (88.5%)
/node/pkg/processor/message.go (0.0%) --> (69.6%)
/node/pkg/processor/observation.go (17.1%) --> (75.0%)
/node/pkg/processor/processor.go (0.0%) --> (37.9%)
/node/pkg/processor/vaa.go (0.0%) --> (72.7%)

/node/pkg/common/grpc.go (56.1%) --> (83.7%)
/node/pkg/common/scissors.go (60.0%) --> (84.0%)
/node/pkg/governor/governor_monitoring.go (7.8%) --> (31.3%)
/node/pkg/node/adminServiceRunnable.go (0.0%) --> (58.6%)
/node/pkg/p2p/p2p.go (56.7%) --> (74.8%)

/node/pkg/publicrpc/publicrpcserver.go (18.0%) --> (31.1%)
/node/pkg/reporter/attestation_events.go (18.8%) --> (43.8%)

/node/pkg/supervisor/supervisor_processor.go (65.6%) --> (83.1%)
/node/pkg/supervisor/supervisor_support.go (0.0%) --> (60.0%)

panoel commented on node/cmd/guardiand/node.go:
```diff
@@ -1353,6 +1352,9 @@ func runNode(cmd *cobra.Command, args []string) {
 	}
 
+	// Finally, add the processor component
+	guardianOptions = append(guardianOptions, node.GuardianOptionProcessor())
```
Help me understand why this is done here instead of just after line 1326.  I don't see any dependencies that would cause it.

tbjump replied:
moved it to line 1327 like you suggested.

panoel commented on node/pkg/node/node.go:
```diff
@@ -84,11 +84,21 @@ func GuardianOptionP2P(p2pKey libp2p_crypto.PrivKey, networkId string, bootstrap
 		}}
 }
 
+func GuardianOptionNoAccountant() *GuardianOption {
```
Did you want to comment that this is a test only function?  Or doesn't it matter?

tbjump replied:
it's not a test-only function; it's a convenience function. I added a docstring comment to this option and while I was at it I added comments to all the other options as well :)

SEJeff commented on node/pkg/node/node.go:
```diff
@@ -244,7 +255,7 @@ func GuardianOptionWatchers(watcherConfigs []watchers.WatcherConfig, ibcWatcherC
 			for _, wc := range watcherConfigs {
 				if _, ok := watchers[wc.GetNetworkID()]; ok {
-					logger.Fatal("NetworkID already configured", zap.String("network_id", string(wc.GetNetworkID())))
+					return fmt.Errorf("NetworkID already configured: %s", string(wc.GetNetworkID()))
```
Minor, but the linter should have caught this. It is idiomatic [to not capitalize the beginning of error strings](https://github.com/golang/go/wiki/CodeReviewComments#error-strings).

tbjump replied:
in this case `NetworkID` is the name of the variable so should be capitalized?

SEJeff replied:
Regardless of the variable name this goes against the [official guidance](https://github.com/golang/go/wiki/CodeReviewComments#error-strings) that error strings. It's a small nit, but FYI and a lot of linters will bomb on this.

tbjump replied:
>>Error strings should not be capitalized (*unless beginning with proper nouns or acronyms*)

SEJeff commented on node/pkg/node/node.go:
```diff
@@ -305,7 +316,7 @@ func GuardianOptionWatchers(watcherConfigs []watchers.WatcherConfig, ibcWatcherC
 					g.runnablesWithScissors["ibcwatch"] = ibc.NewWatcher(ibcWatcherConfig.Websocket, ibcWatcherConfig.Lcd, ibcWatcherConfig.Contract, chainConfig).Run
 				} else {
-					logger.Error("Although IBC is enabled, there are no chains for it to monitor")
+					return errors.New("Although IBC is enabled, there are no chains for it to monitor")
```
```suggestion
					return errors.New("although IBC is enabled, there are no chains for it to monitor")
```

SEJeff commented on node/pkg/node/node.go:
```diff
@@ -335,7 +346,7 @@ func GuardianOptionAdminService(socketPath string, ethRpc *string, ethContract *
 			)
 			if err != nil {
-				logger.Fatal("failed to create admin service socket", zap.Error(err))
+				return err
```
Maybe the previous context was better?

```suggestion
				return fmt.Errorf("failed to create admin service socket: %w", err)
```

SEJeff commented on node/pkg/node/node.go:
```diff
@@ -346,12 +357,12 @@ func GuardianOptionAdminService(socketPath string, ethRpc *string, ethContract *
 			publicrpcUnixService, publicrpcServer, err := publicrpcUnixServiceRunnable(logger, publicGRPCSocketPath, publicRpcLogDetail, g.db, g.gst, g.gov)
 			if err != nil {
-				logger.Fatal("failed to create publicrpc service socket", zap.Error(err))
+				return err
```
Same as before. The context is useful and makes it easier to track down where the error comes from.

SEJeff commented on node/pkg/node/node_test.go:
```diff
@@ -213,6 +233,51 @@ func waitForHeartbeats(t *testing.T, zapObserver *observer.ObservedLogs, gs []*m
 
+func waitForHeartbeatsInMetrics(t *testing.T, ctx context.Context, gs []*mockGuardian) {
+	requests := make([]*http.Request, len(gs))
+	//logger := supervisor.Logger(ctx)
```
```suggestion
```

SEJeff commented on node/pkg/node/node_test.go:
```diff
@@ -213,6 +233,51 @@ func waitForHeartbeats(t *testing.T, zapObserver *observer.ObservedLogs, gs []*m
+				}
+			}
+
+			//logger.Info("node not ready yet", zap.Int("i", i))
```
?

SEJeff commented on node/pkg/node/node_test.go:
```diff
@@ -213,6 +233,51 @@ func waitForHeartbeats(t *testing.T, zapObserver *observer.ObservedLogs, gs []*m
+				continue
+			}
+
+			resp, err := http.DefaultClient.Do(requests[i])
+			if err != nil {
+				resp.Body.Close()
+				continue
+			}
```
What closes the response body here if there is not an error? The general pattern is to defer a resp.Body.Close(), but this is looped so does it make sense to do it at the end or just let them get torn down by the test exiting?

SEJeff commented on node/pkg/node/node_test.go:
```diff
@@ -250,6 +315,66 @@ func someMessage() *common.MessagePublication {
+		bytes := make([]byte, 101)
+		bytes[0] = 1 // tb payload type
+
+		amtBigFloat := big.NewFloat(amtFloat)
+		amtBigFloat = amtBigFloat.Mul(amtBigFloat, big.NewFloat(100000000))
+		amount, _ := amtBigFloat.Int(nil)
+		amtBytes := amount.Bytes()
+		if len(amtBytes) > 32 {
+			panic("amount will not fit in 32 bytes!")
+		}
+		copy(bytes[33-len(amtBytes):33], amtBytes)
+
+		tokenAddr, _ := vaa.StringToAddress(tokenAddrStr)
+		copy(bytes[33:65], tokenAddr.Bytes())
+		binary.BigEndian.PutUint16(bytes[65:67], uint16(tokenChainID))
+		toAddr, _ := vaa.StringToAddress(toAddrStr)
+		copy(bytes[67:99], toAddr.Bytes())
+		binary.BigEndian.PutUint16(bytes[99:101], uint16(toChainID))
+		return bytes
```
Some comments for this would be really nice. It is not intuitive what it is doing if you lack the context of working on the code a bunch.

tbjump commented on node/pkg/node/node_test.go:
```diff
@@ -387,13 +533,19 @@ func TestNodes(t *testing.T) {
+		if WAIT_FOR_LOGS {
+			waitForHeartbeatsInLogs(t, zapObserver, gs)
+		}
 		logger.Info("All Guardians have received at least one heartbeat.")
```
addressed in tbjump/reviewNits

tbjump commented:
Addressed all reviewer feedback. Some nits are fixed in a follow-up PR https://github.com/wormhole-foundation/wormhole/pull/3186 to avoid excessive merge conflicts upstack.

SEJeff commented on node/pkg/node/node_test.go:
```diff
@@ -263,20 +404,27 @@ func makeObsDb(tc []testCase) mock.ObservationDb {
 
 // #nosec G107 -- it's OK to make http requests with `statusAddr` because `statusAddr` is trusted.
 func testStatusServer(ctx context.Context, logger *zap.Logger, statusAddr string) error {
```
Style suggestions for future work like this (not worth blocking this PR whatsoever):
* For functions in tests that have zero assertions, they should not have the name `test` in them
* For functions in test files that aren't tests, please pass `*testing.T` into the function and call [t.Helper()](https://pkg.go.dev/testing#T.Helper) as the first line. This makes stack traces when the tests fails for developers or in CI massively more useful for developers.
